Update documentation on the autoscaling system. (#7126)
knative/serving: docs/scaling/SYSTEM.md

queue them locally, hence the “queue” in its name. The queue-proxy also measures
the incoming request load and reports the **average concurrency** and **requests
per second** on a separate port.

### Autoscaler

The **autoscaler** is a stand-alone pod consisting of three main components:

1. The PodAutoscaler reconciler
2. The **Collector**
3. The **Decider**

The PodAutoscaler reconciler makes sure that any changes to PodAutoscalers (see
API section) are correctly picked up and reflected in either the Decider, the
Collector or both.

The collector is responsible for collecting metrics from the queue-proxies on
the application instances. To do so, it scrapes their internal metrics endpoint
and sums them up to get a metric representative of the entire system. For
scalability, only a sample of all application instances is scraped and the
received metrics are extrapolated to the entire cluster.

The decider gets all metrics available and decides how many pods the
application’s deployment should be scaled to. Basically, the equation to do that
is `want = concurrencyInSystem/targetConcurrencyPerInstance`.

In addition to that, it adjusts the value for a maximum scale up/down rate and

> the min- and max-instances settings on the revision.

Reviewer: I think here's the place to inject "proxy mode for scaled up resource" with some reference to tbc.

### Activator

> The **activator** is a globally shared deployment, that is very scalable. Its

Reviewer: ```suggestion
The **activator** is a globally shared, horizontally scaled deployment. Its
```

Also globally shared becomes semi-truth in the subsetting world, but :man_shrugging:

Author: Well it'll still be shared by all of them logically. Sure, the instances will be scoped down. But the activator as a component is shared between revisions, thus "global".

main purposes are **buffering requests** and **reporting metrics** to the
autoscaler.

The activator is mainly involved around scale to/from zero. When a revision is
scaled to zero instances, the activator will be put into the data path instead
of revision’s instances. If requests would hit this revision, the activator
buffers these requests, pokes the autoscaler with metrics and holds the requests
until instances of the application appear. As soon as that is the case, the
activator will forward the requests it buffered to the new instances while
carefully avoiding to overload the existing instances of the application. The

> activator may also be taken off the data-path for minimal network overhead at

Reviewer: We probably need to reference that it acts as a load balancer.

> activator may also be taken off the data-path for minimal network overhead at

Reviewer: ```suggestion
activator may also be taken on and off the data-path for minimal network overhead at
```

autoscaler. It controls which service to scrape, how to aggregate the data and
more.

Metrics are automatically created from PodAutoscalers by default.

### Decider

**API:** Only exists in-memory today. Only here for completeness sake.

### ServerlessServices (SKS)

**API:** `serverlessservices.networking.internal.knative.dev`

An abstraction on top of Kubernetes Services that controls the data flow and
especially the switch between having the activator as a buffer on the data path
(for example when scaled to zero) or routing to the application instances
directly (when scaled above zero).

This is achieved by creating two Kubernetes services for each revision: A
**public **service and a **private **service.

The private service is a standard Kubernetes service. Its selector is set to
point to the application instances deployed and as the deployment grows and
shrinks, the list of available IPs will do so too.

The public service is an “unmanaged” Kubernetes service. It has no selector and
therefore does not automatically get endpoint management as the private service
does. The public service’s endpoints are managed by the SKS reconciler directly.

> The SKS has two modes: **Proxy** and **Serve**.

Reviewer: ```suggestion
The SKS has two modes: **`Proxy`** and **`Serve`**.
```

> In mode “Serve”, the public service’s endpoints are exactly the same as those of

Reviewer: ```suggestion
In  `Serve` mode, the public service endpoints are exactly the same as those of
```
I think "public service endpoints" is also correct here and generates less stutter.

the private service. All traffic will flow to the revision’s pods.

> In mode Proxy, the public service’s endpoints are the addresses of all

Reviewer: ```suggestion
In `Proxy` mode, the public service endpoints are the addresses of all the
```

Footnote, that it will change soon?

activators in the system. All traffic will flow to the activators.

> ServerlessServices are automatically created from PodAutoscalers by default.

Reviewer: ```suggestion
ServerlessServices are created from PodAutoscalers.
```

into the system, the scraped values will change and the autoscaler will instruct
the revision's deployment to adhere to a given scale.

The SKS keeps track of the changes to the deployment's size through the private
service. It updates the public service accordingly.

### Scaling to zero

![scale-to-0](images/scale-to-0.png)

A revision scales to zero once there are no more requests in the system anymore.
All scrapes from the autoscaler to revision pods return 0 concurrency and the
activator reports the same.

Before actually removing the last pod of the revision the system makes sure that
the activator is in the path and routable. The autoscaler, who decided to
scale-to-zero in the first place, instructs the SKS to use **Proxy** mode, so
all traffic is directed at the activators. The SKS’ public service is now probed
until it has been ensured to return responses from the activator. Once that is
the case and if a grace-period (configurable via _scale-to-zero-grace-period_)
has already passed, the last pod of the revision is removed and the revision has
successfully scaled to zero.

### Scaling from zero

> ![scale-from-0](images/scale-from-0.png)

Reviewer: - 5.1 should be `proxy`?
-

Author: Hmm, I don't think so? It only stays in proxy if TBC is enabled, doesn't it?

If a revision is scaled to zero and a request comes into the system trying to
reach this revision, the system needs to scale it up. As the SKS is in **Proxy**

> mode, the request will reach the activator, which will count it and report it’s

Reviewer: ```suggestion
mode, the request will reach the activator, which will count it and report its
```

appearance to the autoscaler. The activator will then buffer the request and

> watch the SKS’ private service for endpoints to appear.

Reviewer: ```suggestion
watch the SKS’s private service for endpoints to appear.
```

The autoscaler gets the metric from the activator and runs an autoscaling cycle
immediately. That will determine at least one pod to be desired and the

> autoscaler will instruct the revision’s deployment to scale up to N > 0 replica.

Reviewer: ```suggestion
The autoscaler gets the metric from the activator and immediately runs an autoscaling cycle.
That process will determine that at least one pod is desired and the
autoscaler will instruct the revision’s deployment to scale up to N > 0 replicas.
```

The activator eventually sees the endpoints coming up and starts probing it.
Once the probe passes successfully, the respective address will be considered
healthy and used to route the request we buffered to.

> 

Reviewer: ```suggestion
The activator eventually sees the endpoints coming up and starts probing them.
Once the probe passes successfully, the respective address will be considered
healthy and used to route the requests we buffered to.

```
We might have had more than one request at this point.

API section) are correctly picked up and reflected in either the Decider, the
Collector or both.

The collector is responsible for collecting metrics from the queue-proxies on
the application instances. To do so, it scrapes their internal metrics endpoint
and sums them up to get a metric representative of the entire system. For
scalability, only a sample of all application instances is scraped and the
received metrics are extrapolated to the entire cluster.

The decider gets all metrics available and decides how many pods the
application's deployment should be scaled to. Basically, the equation to do that
is `want = concurrencyInSystem/targetConcurrencyPerInstance`.

In addition to that, it adjusts the value for a maximum scale up/down rate and
the min- and max-instances settings on the revision. It also computes how much
burst capacity is left in the current deployment and thus determines whether or
not the activator can be taken off of the data-path or not.

### Activator

The **activator** is a globally shared deployment, that is very scalable. Its
main purposes are **buffering requests** and **reporting metrics** to the
autoscaler.

> The activator is mainly involved around scale to/from zero. When a revision is

Author: I clarified this a bit. We can add references to the TBC document once that exists.

directly (when scaled above zero).

This is achieved by creating two Kubernetes services for each revision: A
**public **service and a **private **service.

The private service is a standard Kubernetes service. Its selector is set to
point to the application instances deployed and as the deployment grows and
shrinks, the list of available IPs will do so too.

The public service is an “unmanaged” Kubernetes service. It has no selector and
therefore does not automatically get endpoint management as the private service
does. The public service's endpoints are managed by the SKS reconciler directly.

The SKS has two modes: **`Proxy`** and **`Serve`**.

In mode `Serve`, the public service endpoints are exactly the same as those of
the private service. All traffic will flow to the revision's pods.

In mode `Proxy`, the public service endpoints are the addresses of all the
activators in the system. All traffic will flow to the activators.

ServerlessServices are created from PodAutoscalers.

## Data flow examples

<!-- Editable versions of the diagrams below are in the Knative shared drive in Scaling/images -->

### Scaling up and down (steady state)

> ![scale-up-down](images/scale-up-down.png)

Reviewer 2: In the image, should the "scrapes" arrow goes to all of the queue proxies?
